File: scripts/token-usage-analyzer.lua
#!/usr/bin/env luajit
--- Analyze Claude Code transcript files to compare token usage with/without codescan tools.
---
--- Parses JSONL transcript files from ~/.claude/projects/*/sessions/ and reports:
--- - Per-tool token usage (input + output)
--- - Codescan MCP tools vs built-in tools (Read/Grep/Glob/Edit)
--- - Session-level summaries
---
--- Usage:
---     luajit scripts/token-usage-analyzer.lua [--last N] [--all]

local cjson = require("cjson")

-- CLI argument parsing
local function parse_args()
	local args = { last = 10 }
	local i = 1
	while i <= #arg do
		if arg[i] == "--last" then
			i = i + 1
			args.last = tonumber(arg[i]) or 10
		elseif arg[i] == "--all" then
			args.all = true
		elseif arg[i] == "--sessions-dir" then
			i = i + 1
			args.sessions_dir = arg[i]
		elseif arg[i] == "--help" or arg[i] == "-h" then
			print("Usage: luajit token-usage-analyzer.lua [--last N] [--all] [--sessions-dir DIR]")
			os.exit(0)
		end
		i = i + 1
	end
	return args
end

-- Find session JSONL files
local function find_session_files(sessions_dir)
	local home = os.getenv("HOME") or "/tmp"
	local base = sessions_dir or (home .. "/.claude/projects")

	local files = {}
	-- Use find to locate JSONL files
	local handle = io.popen('find "' .. base .. '" -name "*.jsonl" -type f 2>/dev/null | head -500')
	if not handle then return files end
	for line in handle:lines() do
		-- Get mtime for sorting
		local mtime_handle = io.popen('stat -f "%m" "' .. line .. '" 2>/dev/null || stat -c "%Y" "' .. line .. '" 2>/dev/null')
		local mtime = 0
		if mtime_handle then
			mtime = tonumber(mtime_handle:read("*l")) or 0
			mtime_handle:close()
		end
		files[#files + 1] = { path = line, mtime = mtime }
	end
	handle:close()

	-- Sort by mtime descending (newest first)
	table.sort(files, function(a, b) return a.mtime > b.mtime end)
	return files
end

-- Parse a single transcript JSONL file
local function parse_transcript(path)
	local tool_stats = {} -- tool_name -> {calls, input_tokens, output_tokens}
	local totals = { input_tokens = 0, output_tokens = 0, turns = 0 }
	local session_date = nil

	local f = io.open(path, "r")
	if not f then return nil end

	for line in f:lines() do
		if line == "" then goto continue end

		local ok, entry = pcall(cjson.decode, line)
		if not ok then goto continue end

		if not session_date and entry.timestamp then
			session_date = entry.timestamp:sub(1, 10)
		end

		if entry.type ~= "assistant" then goto continue end

		local msg = entry.message or {}
		local usage = msg.usage or {}

		local input_tokens = (usage.input_tokens or 0)
			+ (usage.cache_read_input_tokens or 0)
			+ (usage.cache_creation_input_tokens or 0)
		local output_tokens = usage.output_tokens or 0

		totals.input_tokens = totals.input_tokens + input_tokens
		totals.output_tokens = totals.output_tokens + output_tokens
		totals.turns = totals.turns + 1

		-- Extract tool calls
		local content = msg.content or {}
		for _, block in ipairs(content) do
			if block.type == "tool_use" then
				local name = block.name or "unknown"
				if not tool_stats[name] then
					tool_stats[name] = { calls = 0, input_tokens = 0, output_tokens = 0 }
				end
				local ts = tool_stats[name]
				ts.calls = ts.calls + 1
				ts.input_tokens = ts.input_tokens + input_tokens
				ts.output_tokens = ts.output_tokens + output_tokens
			end
		end

		::continue::
	end
	f:close()

	-- Check if any tools were found
	local has_tools = false
	for _ in pairs(tool_stats) do has_tools = true; break end
	if not has_tools then return nil end

	return {
		path = path,
		date = session_date,
		tools = tool_stats,
		totals = totals,
	}
end

-- Tool categorization
local BUILTIN_TOOLS = {
	Read = true, Grep = true, Glob = true, Edit = true, Write = true, Agent = true,
}

local function categorize_tool(name)
	if BUILTIN_TOOLS[name] then return "builtin" end
	if name:match("^mcp__codescan__") then return "codescan" end
	return "other"
end

-- Format number with commas
local function commify(n)
	local s = tostring(math.floor(n))
	local pos = #s % 3
	if pos == 0 then pos = 3 end
	local parts = { s:sub(1, pos) }
	for i = pos + 1, #s, 3 do
		parts[#parts + 1] = s:sub(i, i + 2)
	end
	return table.concat(parts, ",")
end

-- Right-align string to width
local function rpad(s, w)
	s = tostring(s)
	if #s >= w then return s end
	return string.rep(" ", w - #s) .. s
end

-- Left-pad string to width
local function lpad(s, w)
	s = tostring(s)
	if #s >= w then return s end
	return s .. string.rep(" ", w - #s)
end

-- Print report
local function print_report(sessions, last_n)
	if last_n and not sessions.all then
		local limited = {}
		for i = 1, math.min(last_n, #sessions) do
			limited[i] = sessions[i]
		end
		sessions = limited
	end

	print(string.rep("=", 70))
	print(string.format("Claude Code Token Usage Report — %d session(s)", #sessions))
	print(string.rep("=", 70))
	print()

	-- Aggregate
	local agg = {
		builtin = { calls = 0, input_tokens = 0, output_tokens = 0 },
		codescan = { calls = 0, input_tokens = 0, output_tokens = 0 },
		other = { calls = 0, input_tokens = 0, output_tokens = 0 },
	}
	local tool_details = {}
	local total_tokens = 0

	for _, session in ipairs(sessions) do
		for tool_name, stats in pairs(session.tools) do
			local cat = categorize_tool(tool_name)
			local a = agg[cat]
			a.calls = a.calls + stats.calls
			a.input_tokens = a.input_tokens + stats.input_tokens
			a.output_tokens = a.output_tokens + stats.output_tokens

			if not tool_details[tool_name] then
				tool_details[tool_name] = { calls = 0, input_tokens = 0, output_tokens = 0 }
			end
			local td = tool_details[tool_name]
			td.calls = td.calls + stats.calls
			td.input_tokens = td.input_tokens + stats.input_tokens
			td.output_tokens = td.output_tokens + stats.output_tokens
		end
		total_tokens = total_tokens + session.totals.input_tokens + session.totals.output_tokens
	end

	-- Category summary
	print("Category Summary (tokens = input + output attributed to tool-call turns):\n")
	print(string.format("  %s %s %s %s %s",
		lpad("Category", 20), rpad("Calls", 8), rpad("Input Tokens", 14), rpad("Output Tokens", 14), rpad("Total", 14)))
	print(string.format("  %s %s %s %s %s",
		string.rep("-", 20), string.rep("-", 8), string.rep("-", 14), string.rep("-", 14), string.rep("-", 14)))

	for _, pair in ipairs({
		{ "Built-in tools", agg.builtin },
		{ "Codescan MCP", agg.codescan },
		{ "Other", agg.other },
	}) do
		local label, a = pair[1], pair[2]
		local total = a.input_tokens + a.output_tokens
		print(string.format("  %s %s %s %s %s",
			lpad(label, 20), rpad(commify(a.calls), 8),
			rpad(commify(a.input_tokens), 14), rpad(commify(a.output_tokens), 14),
			rpad(commify(total), 14)))
	end
	print(string.format("\n  Total session tokens: %s\n", commify(total_tokens)))

	-- Top tools by call count
	local sorted_tools = {}
	for name, stats in pairs(tool_details) do
		sorted_tools[#sorted_tools + 1] = { name = name, stats = stats }
	end
	table.sort(sorted_tools, function(a, b) return a.stats.calls > b.stats.calls end)

	print("Top tools by call count:\n")
	print(string.format("  %s %s %s %s",
		lpad("Tool", 40), lpad("Cat", 10), rpad("Calls", 8), rpad("Tokens", 14)))
	print(string.format("  %s %s %s %s",
		string.rep("-", 40), string.rep("-", 10), string.rep("-", 8), string.rep("-", 14)))

	for i = 1, math.min(20, #sorted_tools) do
		local t = sorted_tools[i]
		local cat = categorize_tool(t.name)
		local total = t.stats.input_tokens + t.stats.output_tokens
		print(string.format("  %s %s %s %s",
			lpad(t.name, 40), lpad(cat, 10),
			rpad(commify(t.stats.calls), 8), rpad(commify(total), 14)))
	end

	-- Codescan adoption ratios
	local nav_tools = { "Read", "Grep", "Glob" }
	local builtin_nav = 0
	for _, t in ipairs(nav_tools) do
		if tool_details[t] then builtin_nav = builtin_nav + tool_details[t].calls end
	end

	local codescan_nav_tools = {
		"mcp__codescan__search", "mcp__codescan__symbols",
		"mcp__codescan__read_file", "mcp__codescan__query",
	}
	local codescan_nav = 0
	for _, t in ipairs(codescan_nav_tools) do
		if tool_details[t] then codescan_nav = codescan_nav + tool_details[t].calls end
	end

	if builtin_nav + codescan_nav > 0 then
		local ratio = codescan_nav / (builtin_nav + codescan_nav) * 100
		print(string.format("\n  Codescan adoption (navigation): %d/%d calls (%.1f%%)",
			codescan_nav, builtin_nav + codescan_nav, ratio))
	end

	local edit_builtin_tools = { "Edit", "Write" }
	local builtin_edit = 0
	for _, t in ipairs(edit_builtin_tools) do
		if tool_details[t] then builtin_edit = builtin_edit + tool_details[t].calls end
	end

	local codescan_edit_tools = {
		"mcp__codescan__replace_content", "mcp__codescan__replace_symbol",
		"mcp__codescan__replace_lines", "mcp__codescan__insert_at",
		"mcp__codescan__insert_after", "mcp__codescan__insert_before",
		"mcp__codescan__create_file", "mcp__codescan__destroy_file",
	}
	local codescan_edit = 0
	for _, t in ipairs(codescan_edit_tools) do
		if tool_details[t] then codescan_edit = codescan_edit + tool_details[t].calls end
	end

	if builtin_edit + codescan_edit > 0 then
		local ratio = codescan_edit / (builtin_edit + codescan_edit) * 100
		print(string.format("  Codescan adoption (editing):    %d/%d calls (%.1f%%)",
			codescan_edit, builtin_edit + codescan_edit, ratio))
	end

	print()
end

-- Main
local args = parse_args()
local session_files = find_session_files(args.sessions_dir)
io.stderr:write(string.format("Found %d session file(s)\n", #session_files))

local sessions = {}
for _, entry in ipairs(session_files) do
	local session = parse_transcript(entry.path)
	if session then
		sessions[#sessions + 1] = session
	end
end

if #sessions == 0 then
	io.stderr:write("No sessions with tool usage found.\n")
	os.exit(1)
end

local last_n = args.all and #sessions or args.last
print_report(sessions, last_n)

File: scripts/token-usage-analyzer.py
#!/usr/bin/env python3
"""Analyze Claude Code transcript files to compare token usage with/without codescan tools.

Parses JSONL transcript files from ~/.claude/projects/*/sessions/ and reports:
- Per-tool token usage (input + output)
- Codescan MCP tools vs built-in tools (Read/Grep/Glob/Edit)
- Session-level summaries

Usage:
    python3 scripts/token-usage-analyzer.py [--sessions-dir DIR] [--last N]
"""

import json
import os
import sys
import argparse
from pathlib import Path
from collections import defaultdict
from datetime import datetime


def find_session_dirs(base_dir=None):
    """Find all Claude Code session directories."""
    if base_dir:
        return [Path(base_dir)]

    claude_dir = Path.home() / ".claude" / "projects"
    if not claude_dir.exists():
        print(f"No Claude projects found at {claude_dir}", file=sys.stderr)
        return []

    sessions = []
    for project_dir in claude_dir.iterdir():
        if not project_dir.is_dir():
            continue
        # Session files are directly in the project dir
        for f in project_dir.glob("*.jsonl"):
            sessions.append(f)

    return sorted(sessions, key=lambda p: p.stat().st_mtime, reverse=True)


def parse_transcript(path):
    """Parse a JSONL transcript file and extract tool usage with token counts."""
    tool_stats = defaultdict(lambda: {"calls": 0, "input_tokens": 0, "output_tokens": 0})
    session_totals = {"input_tokens": 0, "output_tokens": 0, "turns": 0}
    session_date = None

    try:
        with open(path) as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    entry = json.loads(line)
                except json.JSONDecodeError:
                    continue

                if not session_date and entry.get("timestamp"):
                    session_date = entry["timestamp"][:10]

                msg = entry.get("message", {})
                if entry.get("type") != "assistant":
                    continue

                usage = msg.get("usage", {})
                input_tokens = usage.get("input_tokens", 0) + usage.get("cache_read_input_tokens", 0) + usage.get("cache_creation_input_tokens", 0)
                output_tokens = usage.get("output_tokens", 0)

                session_totals["input_tokens"] += input_tokens
                session_totals["output_tokens"] += output_tokens
                session_totals["turns"] += 1

                # Extract tool calls from content
                content = msg.get("content", [])
                for block in content:
                    if block.get("type") == "tool_use":
                        tool_name = block.get("name", "unknown")
                        tool_stats[tool_name]["calls"] += 1
                        # Attribute this turn's tokens to the tool
                        # (rough — a turn may have multiple tool calls)
                        tool_stats[tool_name]["input_tokens"] += input_tokens
                        tool_stats[tool_name]["output_tokens"] += output_tokens
    except Exception as e:
        print(f"Error parsing {path}: {e}", file=sys.stderr)

    return {
        "path": str(path),
        "date": session_date,
        "tools": dict(tool_stats),
        "totals": session_totals,
    }


BUILTIN_TOOLS = {"Read", "Grep", "Glob", "Edit", "Write", "Agent"}
CODESCAN_TOOLS = {
    "mcp__codescan__search", "mcp__codescan__symbols", "mcp__codescan__read_file",
    "mcp__codescan__replace_content", "mcp__codescan__replace_symbol",
    "mcp__codescan__replace_lines", "mcp__codescan__insert_at",
    "mcp__codescan__insert_after", "mcp__codescan__insert_before",
    "mcp__codescan__create_file", "mcp__codescan__destroy_file",
    "mcp__codescan__diff", "mcp__codescan__references", "mcp__codescan__rename",
    "mcp__codescan__config", "mcp__codescan__status", "mcp__codescan__index",
    "mcp__codescan__query",
}


def categorize_tool(name):
    """Categorize a tool as builtin, codescan, or other."""
    if name in BUILTIN_TOOLS:
        return "builtin"
    if name in CODESCAN_TOOLS or name.startswith("mcp__codescan__"):
        return "codescan"
    return "other"


def print_report(sessions, last_n=None):
    """Print a summary report of token usage."""
    if last_n:
        sessions = sessions[:last_n]

    print(f"\n{'='*70}")
    print(f"Claude Code Token Usage Report — {len(sessions)} session(s)")
    print(f"{'='*70}\n")

    # Aggregate across sessions
    agg_builtin = {"calls": 0, "input_tokens": 0, "output_tokens": 0}
    agg_codescan = {"calls": 0, "input_tokens": 0, "output_tokens": 0}
    agg_other = {"calls": 0, "input_tokens": 0, "output_tokens": 0}
    total_tokens = 0

    tool_details = defaultdict(lambda: {"calls": 0, "input_tokens": 0, "output_tokens": 0})

    for session in sessions:
        for tool_name, stats in session["tools"].items():
            cat = categorize_tool(tool_name)
            target = {"builtin": agg_builtin, "codescan": agg_codescan, "other": agg_other}[cat]
            target["calls"] += stats["calls"]
            target["input_tokens"] += stats["input_tokens"]
            target["output_tokens"] += stats["output_tokens"]

            tool_details[tool_name]["calls"] += stats["calls"]
            tool_details[tool_name]["input_tokens"] += stats["input_tokens"]
            tool_details[tool_name]["output_tokens"] += stats["output_tokens"]

        total_tokens += session["totals"]["input_tokens"] + session["totals"]["output_tokens"]

    # Category summary
    print("Category Summary (tokens = input + output attributed to tool-call turns):\n")
    print(f"  {'Category':<20} {'Calls':>8} {'Input Tokens':>14} {'Output Tokens':>14} {'Total':>14}")
    print(f"  {'-'*20} {'-'*8} {'-'*14} {'-'*14} {'-'*14}")
    for label, agg in [("Built-in tools", agg_builtin), ("Codescan MCP", agg_codescan), ("Other", agg_other)]:
        total = agg["input_tokens"] + agg["output_tokens"]
        print(f"  {label:<20} {agg['calls']:>8} {agg['input_tokens']:>14,} {agg['output_tokens']:>14,} {total:>14,}")

    print(f"\n  Total session tokens: {total_tokens:,}\n")

    # Top tools by call count
    print("Top tools by call count:\n")
    sorted_tools = sorted(tool_details.items(), key=lambda x: x[1]["calls"], reverse=True)
    print(f"  {'Tool':<40} {'Cat':<10} {'Calls':>8} {'Tokens':>14}")
    print(f"  {'-'*40} {'-'*10} {'-'*8} {'-'*14}")
    for tool_name, stats in sorted_tools[:20]:
        cat = categorize_tool(tool_name)
        total = stats["input_tokens"] + stats["output_tokens"]
        print(f"  {tool_name:<40} {cat:<10} {stats['calls']:>8} {total:>14,}")

    # Codescan adoption ratio
    builtin_nav = sum(
        tool_details[t]["calls"]
        for t in ["Read", "Grep", "Glob"]
        if t in tool_details
    )
    codescan_nav = sum(
        stats["calls"]
        for t, stats in tool_details.items()
        if t.startswith("mcp__codescan__") and t in {
            "mcp__codescan__search", "mcp__codescan__symbols",
            "mcp__codescan__read_file", "mcp__codescan__query",
        }
    )

    if builtin_nav + codescan_nav > 0:
        ratio = codescan_nav / (builtin_nav + codescan_nav) * 100
        print(f"\n  Codescan adoption (navigation): {codescan_nav}/{builtin_nav + codescan_nav} calls ({ratio:.1f}%)")

    builtin_edit = sum(
        tool_details[t]["calls"]
        for t in ["Edit", "Write"]
        if t in tool_details
    )
    codescan_edit = sum(
        stats["calls"]
        for t, stats in tool_details.items()
        if t.startswith("mcp__codescan__") and t in {
            "mcp__codescan__replace_content", "mcp__codescan__replace_symbol",
            "mcp__codescan__replace_lines", "mcp__codescan__insert_at",
            "mcp__codescan__insert_after", "mcp__codescan__insert_before",
            "mcp__codescan__create_file", "mcp__codescan__destroy_file",
        }
    )

    if builtin_edit + codescan_edit > 0:
        ratio = codescan_edit / (builtin_edit + codescan_edit) * 100
        print(f"  Codescan adoption (editing):    {codescan_edit}/{builtin_edit + codescan_edit} calls ({ratio:.1f}%)")

    print()


def main():
    parser = argparse.ArgumentParser(description="Analyze Claude Code token usage")
    parser.add_argument("--sessions-dir", help="Directory containing JSONL transcript files")
    parser.add_argument("--last", type=int, default=10, help="Analyze last N sessions (default: 10)")
    parser.add_argument("--all", action="store_true", help="Analyze all sessions")
    args = parser.parse_args()

    session_files = find_session_dirs(args.sessions_dir)
    if not session_files:
        print("No session files found.", file=sys.stderr)
        sys.exit(1)

    print(f"Found {len(session_files)} session file(s)", file=sys.stderr)

    sessions = []
    for path in session_files:
        if path.is_file() and path.suffix == ".jsonl":
            sessions.append(parse_transcript(path))
        elif path.is_dir():
            for f in sorted(path.glob("*.jsonl"), key=lambda p: p.stat().st_mtime, reverse=True):
                sessions.append(parse_transcript(f))

    # Filter to sessions that have tool usage
    sessions = [s for s in sessions if s["tools"]]

    last_n = None if args.all else args.last
    print_report(sessions, last_n)


if __name__ == "__main__":
    main()
